### PR TITLE
fix(seo): make SSR the single owner of JSON-LD on /band/* and /venue/*

### PR DESCRIPTION
**Pull request by BreakableHoodie**
## Summary

`/band/*` and `/venue/*` emitted their structured data **twice** — once server-side, once from the client `<Helmet>` on mount. Same mechanism as the duplicate-canonical bug #784 fixed: Helmet only replaces tags carrying its own `data-rh` marker and appends otherwise. That sweep deliberately scoped itself to `canonical`/`og:*`/`twitter:*`/`description` and left JSON-LD for this change. SSR is now the single owner on both routes, and the client blocks are deleted rather than trimmed.

Closes #790

## What changed

| File | Change |
|------|--------|
| `functions/band/[id].js` | **Fixed the truncation gap first.** `musicGroup.description` reused the 200-char SERP-length value computed for `<meta name="description">`; JSON-LD has no such limit. Added a separate `schemaDesc` at `JSONLD_DESCRIPTION_MAX_LENGTH = 5000`. Meta family unchanged. |
| `frontend/src/pages/BandProfilePage.jsx` | Deleted the client `MusicGroup` block; removed the now-dead `plainDescription` memo and its `stripMarkdownToText` import. Rewrote the ownership comment. |
| `frontend/src/pages/VenuePage.jsx` | Deleted the client `MusicVenue` block. Rewrote the ownership comment (it previously said JSON-LD was out of scope — now false). |
| `functions/utils/ssrMeta.js` | Header comment: JSON-LD ownership follows the same rule, scoped to whether the caller passes a `jsonLd` block. |
| `CLAUDE.md` | Extended the "SSR owns identity meta" section to JSON-LD. |
| `functions/band/__tests__/id.jsonld-description.test.js` | **New.** Meta truncates at 200 while JSON-LD carries the full bio; plus a 5000-char cap boundary case. |
| `frontend/src/pages/__tests__/ssrIdentityMeta.test.jsx` | Extended the `/band/:id` and `/venue/:id` cases to assert exactly one JSON-LD `<script>` per `@type` survives mount. |

## The trap this PR had to avoid

Deleting the client block naively would have been a **regression, not a de-duplication**. The client passed the full untruncated bio; SSR truncated to 200 chars. **44 of the 62 artists with bios exceed 200 characters** (longest 2521), so the schema description would have silently shortened on 44 artist pages. The SSR fix lands first, in the same commit.

Everything else was verified field-by-field as a strict superset before deletion — SSR emits a structured `PostalAddress` where the client emitted a bare string, filters `sameAs` by `http(s)` prefix across all social keys where the client hand-listed eight and safety-checked only Instagram, runs `photo_url` through `normalizeHttpUrl` where the client passed it raw, and adds a `BreadcrumbList` the client never had.

## Sibling sweep

The issue named two pages. There are **five** client-side JSON-LD blocks; the other three are correct as-is and deliberately untouched:

- `ContactPage.jsx`, `AboutPage.jsx` — static-registry routes. `functions/utils/staticPageMeta.js` emits **no** `jsonLd`, so the client block is the only copy. Deleting them would drop the schema entirely — the `og:site_name` trap from #784.
- `EventsPage.jsx` — `/` is deliberately excluded from SSR meta ownership.
- `/event/:slug` — SSR emits `jsonLd` but `App.jsx` has no client block, so nothing to de-duplicate.

## Security / correctness notes

No auth, CSRF, or data-path changes. Net **improvement** to output safety: the deleted client `MusicGroup` emitted `profile.photo_url` and seven of eight social links raw into the DOM; the surviving SSR path filters both. `escapeAttr`/`JSON.stringify` handling in `serveWithInjectedMeta` is unchanged.

## Verification

- [x] Tests: backend 1058 pass / 0 failures, frontend 980 pass / 0 failures
- [x] ESLint: 0 errors both stacks (3 pre-existing warnings, none in touched files)
- [x] Format check: clean both stacks; `markdownlint-cli2 CLAUDE.md` clean
- [x] Build: green (`npm run build --prefix frontend`)
- [x] `validate:openapi`: N/A — `docs/api-spec.yaml` unchanged
- [x] Manual smoke: N/A for rendering; correctness is established by mutation-checked tests instead — see below

**Every new test was mutation-checked, independently re-run by the reviewer:**

| Mutation | Result |
|---|---|
| `schemaDesc` → `plainDesc` in `band/[id].js` | FAILS — `expected 'AAA…'(200) to be 'AAA…'(250)` |
| `JSONLD_DESCRIPTION_MAX_LENGTH` 5000 → 999999 | FAILS — `expected length 5000, got 5001` |
| Restore `VenuePage` client JSON-LD | FAILS — `{ MusicVenue: 2 }` vs `{ MusicVenue: 1 }` |
| Restore `BandProfilePage` client JSON-LD | FAILS — `{ MusicGroup: 2 }` vs `{ MusicGroup: 1 }` |

## Notes for reviewer

**React 19 + `react-helmet-async` script placement — found while mutation-checking.** The first version of the frontend test passed against the *unfixed* code. Root cause: under React 19, helmet-async v3 only auto-hoists `<script async>` into `<head>`; a bare `<script type="application/ld+json">` renders in place in the React tree, i.e. inside `<body>`. Scoping the count to `document.head` therefore observed nothing either way. The test now queries the whole document, and the reasoning is recorded in a comment above `countJsonLdByType()` so the next person doesn't rediscover it. This is orthogonal to #790 but affects how any future test asserts on client-rendered JSON-LD. (JSON-LD in `<body>` is valid — Google parses it anywhere in the document — so this is a testing concern, not a live SEO bug.)

**`stripMarkdownToText` is now an export with no production consumers** — this PR removed its last caller, and the plain-text path it served is now server-side (`toPlainText` in `ssrMeta.js`). Left in place deliberately: it is small, tested, and its tests cover a security-relevant `DOMPurify` `ALLOWED_TAGS: []` configuration. Deleting it plus ~40 lines of tests is a utils cleanup, not part of a JSON-LD de-duplication. Flagging it rather than doing it silently — happy to remove if you'd rather.

**One CodeRabbit finding rejected**, filed as #797 instead: it wanted `<meta name="keywords">` removed from `BandProfilePage` on the grounds that client Helmet "must own only `<title>`" on SSR routes. That reasoning is wrong — no SSR handler emits `keywords`, so removing it drops the tag rather than de-duplicating it, which is precisely the `og:site_name` trap. The conclusion has independent merit (Google has ignored `keywords` since 2009), so it's tracked separately rather than folded into a de-duplication diff.

---
Built by Sonny · Reviewed by Theo · 🤖 [Claude Code](https://claude.com/claude-code)

## Summary by CodeRabbit

* **SEO & Metadata**
  * Improved server-rendered structured data for band and venue pages.
  * Prevented duplicate JSON-LD entries when pages load in the browser.
  * Preserved longer band biographies in structured data while maintaining concise search and social descriptions.
  * Limited band structured-data descriptions to 5,000 characters and meta/social descriptions to 200 characters.

* **Tests**
  * Added validation for metadata ownership and exactly one structured-data entry per schema type.

### CODE REVIEW

File: functions/band/[id].js [hunk 1]
```diff
@@ -36,7 +36,14 @@ export async function onRequest(context) {
...
   // og:image/twitter:image back to the branded default (#644 review, CodeRabbit).
   const safePhotoUrl = normalizeHttpUrl(band.photo_url);
+  // 200 is a SERP-friendly length for the meta/og/twitter description family
+  // below. JSON-LD's musicGroup.description has no such limit -- truncating
+  // it there is a real content loss, not a de-duplication (#790): 44 of 62
+  // artist bios in production run past 200 chars, longest 2521. 5000 gives
+  // headroom over that without an unbounded payload.
+  const JSONLD_DESCRIPTION_MAX_LENGTH = 5000;
   const plainDesc = toPlainText(band.description, 200);
+  const schemaDesc = toPlainText(band.description, JSONLD_DESCRIPTION_MAX_LENGTH);
```
[the remark this answers is not in the document]
Comment (BreakableHoodie) [reply functions/band/[id].js:52]: Fixed in ede81fa. Extracted `truncatePlainText()` from `toPlainText`'s tail so the strip work happens once and the two lengths are derived from the same result:

```js
const schemaDesc = toPlainText(band.description, JSONLD_DESCRIPTION_MAX_LENGTH);
const plainDesc = truncatePlainText(schemaDesc, 200);
```

Verified this is byte-identical to the previous direct `toPlainText(desc, 200)` across 13 inputs — markdown, HTML entities, and the 199/200/201/4999/5000/5001 boundaries. It holds because the shorter cut is always a prefix of the longer one. Both mutation checks still fail correctly on the new code path.

File: functions/band/__tests__/id.jsonld-description.test.js [hunk 2]
```diff
@@ -0,0 +1,104 @@
...
+
+    // musicGroup is the first schema in jsonLd: [musicGroup, breadcrumb], so
+    // the first <script type="application/ld+json"> block in document order
+    // is always musicGroup -- same assumption the #644 og-image test makes.
+    const jsonLdMatch = html.match(/<script type="application\/ld\+json">([\s\S]*?)<\/script>/);
+    expect(jsonLdMatch).not.toBeNull();
+    const musicGroup = JSON.parse(jsonLdMatch[1]);
+    expect(musicGroup["@type"]).toBe("MusicGroup");
+    expect(musicGroup.description).toBe(longBio);
+    expect(musicGroup.description.length).toBe(250);
```
[the remark this answers is not in the document]
Comment (BreakableHoodie) [reply functions/band/__tests__/id.jsonld-description.test.js]: Fixed in ede81fa. Added a `findSchema(html, type)` helper; all three assertions now select by `@type` rather than by position, with a comment recording why. Agreed on the reasoning — the route emits `jsonLd: [musicGroup, breadcrumb]` so index 0 works today, but that would break silently the moment the route or `index.html` gained another block.

File: frontend/src/pages/__tests__/ssrIdentityMeta.test.jsx [hunk 3]
```diff
@@ -80,12 +85,58 @@ function countIdentityTags() {
...
+// a bare <script type="application/ld+json"> (no `async`, as this codebase's
+// JSON-LD blocks are) renders in place in the React tree instead, i.e.
+// inside <body>, wherever <Helmet> sits as a component. A client-duplicated
+// JSON-LD block would therefore land in <body>, NOT <head> -- confirmed by
+// instrumenting react-helmet-async's React19Dispatcher directly. Scoping
+// this count to document.head would make the assertion below pass whether or
+// not the client still declares the block, which is the vacuous-test failure
+// mode this repo has hit before (see CLAUDE.md "Vacuous test defect class"
+// memory) -- querying the whole document is the only selector that can
+// actually observe a real duplicate here.
```
[the remark this answers is not in the document]
Comment (BreakableHoodie) [reply frontend/src/pages/__tests__/ssrIdentityMeta.test.jsx:120]: Correct — verified there is no "Vacuous test defect class" section in `CLAUDE.md`; that came from a private note and should not have been cited as if it were repo documentation. Fixed in ede81fa by keeping the rationale and dropping the phantom reference:

> a test that survives both the correct and the broken implementation, i.e. no guard at all. This was caught by mutation-checking (restoring the deleted client block and confirming the test then fails)

File: CLAUDE.md [hunk 4]
```diff
@@ -194,14 +194,14 @@ Use `document.title = pageTitle` directly in a `useEffect` within the page compo
 
 Example: `frontend/src/pages/BandProfilePage.jsx` — uses both `<Helmet>` (for other meta) and `document.title = ...` for the title.
 
-### SSR owns identity meta; client `<Helmet>` owns only `<title>` (+ JSON-LD) — never canonical/og:*/twitter:*/description
+### SSR owns identity meta and JSON-LD where it emits either; client `<Helmet>` owns only `<title>` — never canonical/`og:*`/`twitter:*`/description, and never JSON-LD on a route whose SSR handler already emits it
```
[the remark this answers is not in the document]
Comment (BreakableHoodie) [reply CLAUDE.md]: Good catch, and it lands harder than stated — the wording contradicted the reasoning I used to **reject** your earlier finding on this same PR (that `<meta name="keywords">` must stay because no SSR handler emits it, filed separately as #797). Fixed in ede81fa at both sites.

`CLAUDE.md:197` now reads "the client `<Helmet>` owns only what SSR does not emit", and the section gained an explicit paragraph:

> **The rule is "SSR owns what SSR emits," not "the client may only own `<title>`."** A tag no SSR handler emits stays client-owned and must not be deleted in an ownership sweep — deleting it drops the tag rather than de-duplicating it.

`ssrIdentityMeta.test.jsx` header updated to match, with the `keywords` case named as the live example.

File: functions/utils/ssrMeta.js [hunk 5]
```diff
@@ -86,7 +95,17 @@ export function toPlainText(rawInput, maxLength = 200) {
   };
   text = text.replace(/&(?:nbsp|amp|lt|gt|quot|#39);/gi, (m) => ENTITY_MAP[m.toLowerCase()] ?? m);
 
-  text = text.replace(/\s+/g, " ").trim();
+  return truncatePlainText(text.replace(/\s+/g, " ").trim(), maxLength);
+}
```
[the remark this answers is not in the document]
Comment (BreakableHoodie) [reply functions/utils/ssrMeta.js:99]: Confirmed and fixed in 0adb148 — but clamped rather than thrown.

You are right that `truncatePlainText("abc", 0)` returned `"ab…"`. Worth noting it is also **pre-existing**: the same arithmetic lived inside `toPlainText` before this PR split the function out, so `toPlainText("abc", 0)` had the identical defect. Fixing it in the extracted helper fixes both call paths at once.

I did not take the `throw new RangeError(...)` though. `functions/band/[id].js` calls both `toPlainText` and `truncatePlainText` at lines 52/55, which are **outside** either of that handler's `try` blocks (line 25 wraps the D1 lookup, line 83 the `social_links` parse). A throw there escapes `onRequest` and the Pages runtime returns a 500 — whereas this module's header explicitly promises the opposite:

> On ANY problem (gate closed, not found, DB error, asset fetch fail) the caller falls back to `env.ASSETS.fetch(request)` — the SPA still renders client-side, just without the rich tags.

Trading a slightly-too-long string for a dead page would invert that contract. So:

```js
if (!(maxLength > 0)) return "";
```

which also absorbs `undefined` and `NaN` without a separate branch.

Tests added per your request — `0`, negative, `undefined`, `NaN`, plus the degenerate `maxLength: 1` (budget goes to the ellipsis) and the re-truncation equivalence the band route depends on. Mutation-checked: removing the clamp fails with `expected 'ab…' to be ''`.
